machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MCOConfig describes configuration for MachineConfigOperator.
+type MCOConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +required
+	Spec MCOConfigSpec `json:"spec"`
+}
+
+// MCOConfigSpec is the spec for MCOConfig resource.
+type MCOConfigSpec struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MCOConfigList is a list of MCOConfig resources
+type MCOConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MCOConfig `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfig describes configuration for MachineConfigController.
+// This is currently only used to drive the machineconfigs generated by the TemplateController.
+type ControllerConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControllerConfigSpec   `json:"spec"`
```

Comment:
do we maybe need a `+required` on this?

Reply:
Depends on the tool used to generate OpenAPI. For controller-gen from kubebuilder "required" is the default. Our github.com/openshift/crd-schema-gen fork has optional as default. Future kubebuilder will support both cases.